### PR TITLE
Added some engineering design docs.

### PR DESCRIPTION
**Pull request by OmarIthawi**
This is the most basic engieering docs I could think of. 

I have no plans to extend it in the future, unless we get some repeated questions that justifies spending time on it.

After it's merged these docs would make it into: https://course-access-groups.readthedocs.io/

### TODO
 - [ ] Wait up to March 27th to get a review from John on the engineering design docs.

### CODE REVIEW

**Comment by OmarIthawi** [comment]
@johnbaldwin Please review this PR at your leisure. Being a doc-only PR allow us to keep it open for a longer time so we don't rush it.

File: docs/engineering_design.rst [hunk 1]
```diff
@@ -3,8 +3,146 @@
...
+
+
+Future Plans
+------------
+
+This application works on a *modified* fork of the Hawthorn release. This
+section denotes plans to support future releases the impact on the
+architecture of this app.
+
+Supporting Bridgkeeper in Juniper
```
Comment (thraxil) [inline docs/engineering_design.rst]:
```suggestion
Supporting Bridgekeeper in Juniper
```

Not familiar with it, but *assuming* that's a typo.

Comment (OmarIthawi) [reply docs/engineering_design.rst]: It is :)

File: docs/engineering_design.rst [hunk 2]
```diff
@@ -3,8 +3,149 @@
...
+
+ * **Course:** Courses that belongs to this group. To simplify the use
+   relation with the Django REST Framewor it has been moved to its
+   own model named ``GroupCourse``.
+
+ * **User:** Users who are member of this group. For the same reasons above
+   it has it's own model named ``Membership``.
+
+This model is the only essential model for this app to run. The others can
+be thought as complementary.
```
Comment (johnbaldwin) [inline docs/engineering_design.rst]: A tiny suggestion: "The others can be thought **of** as complementary"

File: docs/engineering_design.rst [hunk 3]
```diff
@@ -3,8 +3,149 @@
...
+Therefore the app was designed to be somewhat simple with a couple of
+hard-coded assumptions about the platform that would uses it.
+
+While these assumptions are hard-coded in the initial release, some of them
+will be revisited in future release to be either modified or completely
+removed.
+
+ * **Multi-tenancy:** One of the main assumptions is that the platform
+   uses the multi-tenancy (Django Site) features without
+   having a main site of its own.
```
Comment (johnbaldwin) [inline docs/engineering_design.rst]: can "having a main" be added to the previous line?

File: docs/engineering_design.rst [hunk 4]
```diff
@@ -3,8 +3,149 @@
...
+In addition to this documentation please consider taking a look at the
+`Access Control Backends thread`_ on Open edX Discuss.
+It touches on a couple of related topics regarding Bridgekeeper and some
+recommendation from the edX engineers.
+
+Single-site Setups
+~~~~~~~~~~~~~~~~~~
+The majority of the Open edX Platform installations are single-site setups
+in which Site Configurations isn't used. This application doesn't support
+such installations at the moment. In order to support this installation
```
Comment (johnbaldwin) [inline docs/engineering_design.rst]: suggest a comma after "installation" or rephrase to "Several modifications need to be done to support this installation."

File: docs/engineering_design.rst [hunk 5]
```diff
@@ -3,8 +3,149 @@
...
+ - A new flag to select the mode of the app e.g.
+   ``FEATURES['COURSE_ACCESS_GROUPS_IS_MULTISITE']``.
+
+ - ``CourseAccessGroup.organization`` to be optional: This can be done in two
+   methods: A) Make the field ``null=True`` or make a new
+   ``SiteWideCourseAccessGroup`` to avoid having null values. My (Omar)
+   preference is having a second module.
+
+ - All queries for both ``UserOrganizationMapping`` and
+   ``OrganizationCourse`` should be placed behind the feature flag.
```
Comment (johnbaldwin) [inline docs/engineering_design.rst]: What does "placed behind the feature flag" mean?

File: docs/engineering_design.rst [hunk 6]
```diff
@@ -3,8 +3,149 @@
...
+
+
+**Open Question:** Do we want to support both site-wide and
+organization-specific mode at the same time? Maybe not.
```
Comment (johnbaldwin) [inline docs/engineering_design.rst]: Maybe remove the "Maybe not" or a brief explanation why we don't want to support both.

**Pull request merged**